crates/turbopack-dev-server/src/source/mod.rs

```diff
@@ -37,32 +37,58 @@ pub struct ProxyResult {
 /// attached and when combining results this specificity should be used to order
 /// results.
 #[turbo_tasks::value(shared)]
-pub struct ContentSourceResult {
-    pub specificity: SpecificityVc,
-    pub content: ContentSourceContentVc,
+pub enum ContentSourceResult {
+    NotFound,
+    NeedData(NeededData),
+    Result {
+        specificity: SpecificityVc,
+        get_content: GetContentSourceContentVc,
+    },
+}
+
+#[turbo_tasks::value_impl]
+impl ContentSource for ContentSourceResult {
+    #[turbo_tasks::function]
+    fn get(
+        self_vc: ContentSourceResultVc,
+        _path: &str,
+        _data: Value<ContentSourceData>,
+    ) -> ContentSourceResultVc {
+        self_vc
+    }
 }
 
 #[turbo_tasks::value_impl]
 impl ContentSourceResultVc {
     /// Wraps some content source content with exact match specificity.
     #[turbo_tasks::function]
-    pub fn exact(content: ContentSourceContentVc) -> ContentSourceResultVc {
-        ContentSourceResult {
+    pub fn exact(get_content: GetContentSourceContentVc) -> ContentSourceResultVc {
+        ContentSourceResult::Result {
             specificity: SpecificityVc::exact(),
-            content,
+            get_content,
         }
         .cell()
     }
 
+    /// Wraps some content source content with exact match specificity.
+    #[turbo_tasks::function]
+    pub fn need_data(data: Value<NeededData>) -> ContentSourceResultVc {
+        ContentSourceResult::NeedData(data.into_value()).cell()
+    }
+
     /// Result when no match was found with the lowest specificity.
     #[turbo_tasks::function]
     pub fn not_found() -> ContentSourceResultVc {
-        ContentSourceResult {
-            specificity: SpecificityVc::not_found(),
-            content: ContentSourceContent::NotFound.cell(),
-        }
-        .cell()
+        ContentSourceResult::NotFound.cell()
+    }
+}
+
+#[turbo_tasks::value_trait]
+pub trait GetContentSourceContent {
```

Comment:
Please document this trait and how it interacts with `ContentSource`. I'm having trouble understanding why we need the additional indirection.

From looking at the trait definition, rather than a `GetContentSourceContent`, this feels more like a `DeferredContent`, i.e. a content source's content that we know exists but is yet to be computed.